crates/napi/src/threadsafe_function.rs

```diff
@@ -231,39 +234,48 @@ impl<T: 'static, ES: ErrorStrategy::T> ThreadsafeFunction<T, ES> {
   ///
   /// "ref" is a keyword so that we use "refer" here.
   pub fn refer(&mut self, env: &Env) -> Result<()> {
-    if self.aborted.load(Ordering::Acquire) {
+    let is_aborted = self.aborted.lock().unwrap();
+    if *is_aborted {
       return Err(Error::new(
         Status::Closing,
         "Can not ref, Thread safe function already aborted".to_string(),
       ));
     }
+    drop(is_aborted);
+    self.ref_count.fetch_add(1, Ordering::AcqRel);
```

Comment:
I don't quite understand ref counter usage in `refer`/`unref`. Original napi calls are not referenced counted and idempotent. And they don't control memory management, `release`/`acquire` do. So, while this fixes QueryEngine problem, doesn't it also introduce leak in other direction? For example:

```rust
let mut tsfn: ThreadsafeFunction<String> =
    func.create_threadsafe_function(0,...)

  tsfn.refer(&env)?;
``` 

This `refer` call is perfectly safe when using underlying APIs, albeit a little bit useless, but now it will cause function to leak.

Reply:
The `refer/unref` API should not exist at all, it should be exposed as `clone/drop` in `Rust` world. 🤔 Maybe we should remove it in next major release.

> This refer call is perfectly safe when using underlying APIs, albeit a little bit useless, but now it will cause function to leak.

Nothing leak after this PR, the `Arc` will be dropped in the `fn drop` in `ThreadsafeFunction`

Reply:
Sorry, i think you are mixing up `ref`/`unref` and `acquire`/`release` API. 
`refer`/`unref` do not have anything to do with threadsafe function reference counting, despite what their name might suggest. They solely exist for letting or forbidding llibuv event loop completion without function cleanup. No matter how many times you call `napi_unref_threadsafe_function` it won't release the function and no matter how many times you call `napi_ref_threadsafe_function` it won't prevent function from being released. Moreover, no matter how many times you'd call  `napi_ref_threadsafe_function`, single `napi_unref_threadsafe_function` call will undo everything.  

So that's why adding reference counter in Rust wrapper does not make sense to me — this is not how N-API calls you are wrapping work.


Now let's walk through my example:
1. New function is created, internal reference counter is at 1.
2. We call `refer`. Original C-API does not do anything since function is already referenced. Rust wrapper on the other hand increments internal counter by 1, it is now at 2.
3. We drop the function. Rust internal ref counter is at 2. We skip release call, decrement the counter and finish the drop. Now neither the function is released, nor we have any pointer to the function. Function and it's context is leaked.